app/channel.go

```diff
@@ -45,71 +45,60 @@ func (a *App) JoinDefaultChannels(teamId string, user *model.User, shouldBeAdmin
-		}
-
-		if *a.Config().ServiceSettings.ExperimentalEnableDefaultChannelLeaveJoinMessages {
-			if requestor == nil {
-				if err := a.postJoinTeamMessage(user, townSquare); err != nil {
-					mlog.Error(fmt.Sprint("Failed to post join/leave message", err))
-				}
-			} else {
-				if err := a.postAddToTeamMessage(requestor, user, townSquare, ""); err != nil {
-					mlog.Error(fmt.Sprint("Failed to post join/leave message", err))
-				}
+	if len(a.Config().TeamSettings.ExperimentalDefaultChannels) == 0 {
+		defaultChannelList = append(defaultChannelList, "off-topic")
+	} else {
+		seenChannels := map[string]bool{}
+		for _, channelName := range a.Config().TeamSettings.ExperimentalDefaultChannels {
+			if seenChannels[channelName] != true {
+				defaultChannelList = append(defaultChannelList, channelName)
+				seenChannels[channelName] = true
 			}
 		}
-
-		a.InvalidateCacheForChannelMembers(result.Data.(*model.Channel).Id)
 	}
 
-	if result := <-a.Srv.Store.Channel().GetByName(teamId, "off-topic", true); result.Err != nil {
-		err = result.Err
-	} else if offTopic := result.Data.(*model.Channel); offTopic.Type == model.CHANNEL_OPEN {
+	for _, channelName := range defaultChannelList {
+		if result := <-a.Srv.Store.Channel().GetByName(teamId, channelName, true); result.Err != nil {
+			err = result.Err
+		} else {
 
-		cm := &model.ChannelMember{
-			ChannelId:   offTopic.Id,
-			UserId:      user.Id,
-			SchemeUser:  true,
-			SchemeAdmin: shouldBeAdmin,
-			NotifyProps: model.GetDefaultChannelNotifyProps(),
-		}
+			channel := result.Data.(*model.Channel)
 
-		if cmResult := <-a.Srv.Store.Channel().SaveMember(cm); cmResult.Err != nil {
-			err = cmResult.Err
-		}
-		if result := <-a.Srv.Store.ChannelMemberHistory().LogJoinEvent(user.Id, offTopic.Id, model.GetMillis()); result.Err != nil {
-			mlog.Warn(fmt.Sprintf("Failed to update ChannelMemberHistory table %v", result.Err))
-		}
+			if channel.Type != model.CHANNEL_OPEN {
+				continue
+			}
 
-		if requestor == nil {
-			if err := a.postJoinChannelMessage(user, offTopic); err != nil {
-				mlog.Error(fmt.Sprint("Failed to post join/leave message", err))
+			cm := &model.ChannelMember{
+				ChannelId:   channel.Id,
+				UserId:      user.Id,
+				SchemeUser:  true,
+				SchemeAdmin: shouldBeAdmin,
+				NotifyProps: model.GetDefaultChannelNotifyProps(),
 			}
-		} else {
-			if err := a.PostAddToChannelMessage(requestor, user, offTopic, ""); err != nil {
-				mlog.Error(fmt.Sprint("Failed to post join/leave message", err))
+
+			if cmResult := <-a.Srv.Store.Channel().SaveMember(cm); cmResult.Err != nil {
+				err = cmResult.Err
+			}
+			if result := <-a.Srv.Store.ChannelMemberHistory().LogJoinEvent(user.Id, channel.Id, model.GetMillis()); result.Err != nil {
+				mlog.Warn(fmt.Sprintf("Failed to update ChannelMemberHistory table %v", result.Err))
+			}
+
+			if channel.Name == model.DEFAULT_CHANNEL && *a.Config().ServiceSettings.ExperimentalEnableDefaultChannelLeaveJoinMessages {
```

Comment:
I think we still need to post the "joined the channel" messages for other channels that aren't town square

Reply:
@hmhealey The setting is named as `ExperimentalEnableDefaultChannelLeaveJoinMessages` though, shouldn't it apply to all default channels?

Reply:
If it's a default channel, wouldn't `joined the team` still be correct, and not `joined the channel`, considering they just joined the team and thus ended up automatically instead of intentionally joining the channel? That would - for me - basically mean revert the latest change and remove from L88 the `channel.Name == model.DEFAULT_CHANNEL` condition, all other behaviors would appear inconsistent for me:

1. The setting is named `EnableDefaultChannelLeaveJoinMessage`, it should control all join/leave message for every default channel
2. If town square uses the `joined the team` wording, why use `joined the channel` for other channels? If you stick to joined the team, you can differ between team joins and intentional channel joins
3. If you dont send anything for default channels not being town square (latest PR state), people won't see new members joining.

I don't really have a particular strong on any cases, but posting `joined the team` to all default channels controllable by the experimental var would make the most sense to me.

Reply:
1 - 

> If it's a default channel, wouldn't `joined the team` still be correct, and not `joined the channel`

I'm not sure, but that's beyond the scope of this PR. Right now Off-Topic gets `joined the channel` message and we should keep it, though it can be discussed if a separate PR is worth opening.

PS: Town Square uses the `joined the team` wording because it's the first channel you view by default and it's the only channel you cannot leave.

2 - `EnableDefaultChannelLeaveJoinMessage`: As per the name, it should apply to default channels. If we only want it to apply to Town Square, we can change the setting name to `EnableTownSquareLeaveJoinMessage`. 

My two cents: I think keeping it as `EnableDefaultChannelLeaveJoinMessage` is preferred as it was added for performance reasons where adding multiple users at once triggering channel join/leave messages was causing issues due to the large number of database writes

Reply:
The current behaviour of that setting actually only disables the messages for the Town Square, not the Off Topic channel.

I don't have any strong opinions on whether or not we change the functionality of the setting or the "joined the channel" vs "joined the team" messages. I'd be fine changing both of those as long as it's a conscious decision and not just something we forgot about

Reply:
@DSchalla If you're okay let's go with my proposal above?